google/cloud/storage/client.cc

```diff
@@ -105,6 +105,44 @@ integrity checks using the DisableMD5Hash() and DisableCrc32cChecksum() options.
   return metadata;
 }
 
+void Client::DownloadFileImpl(internal::ReadObjectRangeRequest const& request,
+                              std::string const& file_name) {
+  std::unique_ptr<internal::ObjectReadStreambuf> streambuf =
+      raw_client_->ReadObject(request).second;
+  ObjectReadStream stream(std::move(streambuf));
+  if (stream.eof() and not stream.IsOpen()) {
+    std::string msg = __func__;
+    msg += ": cannot open source object ";
+    msg += request.object_name();
+    msg += " in bucket ";
+    msg += request.bucket_name();
+    google::cloud::internal::RaiseRuntimeError(msg);
+  }
+  std::ofstream os(file_name);
+  if (not os.is_open()) {
+    std::string msg = __func__;
+    msg += ": cannot open destination file ";
+    msg += file_name;
+    google::cloud::internal::RaiseRuntimeError(msg);
+  }
+  std::string buffer;
```

Comment:
Clarifying question:
I notice the library uses `std::string` versus `std::vector<uint8_t>`, I've always wondered why this was the case. It's more for learning and not a requested to change.

Reply:
The I/O APIs (iostreams, FILE*, libcurl, etc) typically use `char*`, so using `std::vector<std::uint8_t>` would require casting, while `std::string` does not.  For what is worth, proto3 also uses `std::string` to represent `bytes` in C++:

https://developers.google.com/protocol-buffers/docs/reference/cpp-generated#proto3_string

So I probably did not think too much about the choice for "How do we represent raw bytes in the library" and just went with the familiar `std::string`.

Reply:
Thanks for the info both! W.r.t Matt's comment "semantically for displaying text", that's one reason I was a bit confused by using `std::string` versus `std::vector<std::uint8_t>`. In this case std::string is just being used as a buffer for arbitrary bytes. 

It's good to know that gRPC does the same thing at least for consistency. Looks like a similar discussion has occurred on [SO](https://stackoverflow.com/questions/1556229/vector-unsigned-char-vs-string-for-binary-data). Thanks again.